File: Src/Shared/Models/Project.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json;

namespace ProjectDashboard.Shared.Models;

public class Project
{
    [Key]
    public int Id { get; set; }
    
    [Required]
    public string Name { get; set; } = string.Empty;
    
    [Required]
    public string Path { get; set; } = string.Empty;
    
    public DateTime? LastCommit { get; set; }
    
    public string? LastCommitMessage { get; set; }
    
    public string? CurrentBranch { get; set; }
    
    public bool IsPinned { get; set; }
    
    public DateTime? ScannedAt { get; set; }

    
    public bool IsForked { get; set; }
    public string? RemoteUrl { get; set; }
    public string? UpstreamUrl { get; set; }

    
    public string Tags { get; set; } = "";          
    public string Notes { get; set; } = "";          
    public string Goals { get; set; } = "[]";        

    
    public string MetadataJson { get; set; } = "{}";

    [NotMapped]
    public Dictionary<string, string> Metadata
    {
        get => JsonSerializer.Deserialize<Dictionary<string, string>>(MetadataJson) ?? new();
        set => MetadataJson = JsonSerializer.Serialize(value);
    }

    
    [NotMapped]
    public List<GoalItem> GoalItems
    {
        get
        {
            try { return JsonSerializer.Deserialize<List<GoalItem>>(Goals) ?? new(); }
            catch { return new(); }
        }
        set => Goals = JsonSerializer.Serialize(value);
    }

    
    [NotMapped]
    public string[] TagList => string.IsNullOrWhiteSpace(Tags)
        ? Array.Empty<string>()
        : Tags.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

    

    [NotMapped]
    public string Status
    {
        get
        {
            if (LastCommit == null) return "Unknown";
            var days = (DateTime.UtcNow - LastCommit.Value).TotalDays;
            if (days <= 7) return "Active";
            if (days <= 30) return "Recent";
            if (days <= 90) return "Stalled";
            return "Archived";
        }
    }

    [NotMapped]
    public string TimeAgo
    {
        get
        {
            if (LastCommit == null) return "never";
            var span = DateTime.UtcNow - LastCommit.Value;
            if (span.TotalMinutes < 1) return "just now";
            if (span.TotalHours < 1) return $"{(int)span.TotalMinutes}m ago";
            if (span.TotalDays < 1) return $"{(int)span.TotalHours}h ago";
            if (span.TotalDays < 7) return $"{(int)span.TotalDays}d ago";
            if (span.TotalDays < 30) return $"{(int)(span.TotalDays / 7)} week{((int)(span.TotalDays / 7) == 1 ? "" : "s")} ago";
            if (span.TotalDays < 365) return $"{(int)(span.TotalDays / 30)} month{((int)(span.TotalDays / 30) == 1 ? "" : "s")} ago";
            return $"{(int)(span.TotalDays / 365)} year{((int)(span.TotalDays / 365) == 1 ? "" : "s")} ago";
        }
    }

    [NotMapped]
    public string[] TechPills
    {
        get
        {
            var pills = new List<string>();
            var meta = Metadata;
            if (meta.TryGetValue("techs", out var techs) && !string.IsNullOrWhiteSpace(techs))
            {
                pills.AddRange(techs.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries));
            }
            else if (meta.TryGetValue("inferred_type", out var inferredType) && inferredType != "Unknown")
            {
                pills.Add(inferredType);
            }
            return pills.ToArray();
        }
    }

    
    [NotMapped]
    public int ContributorCount
    {
        get
        {
            var meta = Metadata;
            return meta.TryGetValue("contributors", out var c) && int.TryParse(c, out var n) ? n : 0;
        }
    }

    [NotMapped]
    public int TotalCommits
    {
        get
        {
            var meta = Metadata;
            return meta.TryGetValue("total_commits", out var c) && int.TryParse(c, out var n) ? n : 0;
        }
    }

    [NotMapped]
    public int BranchCount
    {
        get
        {
            var meta = Metadata;
            return meta.TryGetValue("branch_count", out var c) && int.TryParse(c, out var n) ? n : 0;
        }
    }

    [NotMapped]
    public int DocCount
    {
        get
        {
            var meta = Metadata;
            return meta.TryGetValue("doc_count", out var c) && int.TryParse(c, out var n) ? n : 0;
        }
    }

    [NotMapped]
    public string OwnershipLabel => IsForked ? "Fork" : "Owned";

    [NotMapped]
    public string Initial => !string.IsNullOrEmpty(Name) ? Name[0].ToString().ToUpper() : "?";
}

public class GoalItem
{
    public string Text { get; set; } = "";
    public bool Done { get; set; }
}

File: Src/WebRails/app/models/project.rb
class Project < ApplicationRecord
  validates :name, presence: true
  validates :path, presence: true, uniqueness: { case_sensitive: false }

  def metadata
    JSON.parse(metadata_json.presence || "{}")
  rescue JSON::ParserError
    {}
  end

  def metadata=(hash)
    self.metadata_json = hash.to_json
  end

  def tag_list
    return [] if tags.blank?
    tags.split(",").map(&:strip).reject(&:blank?)
  end

  def goal_items
    JSON.parse(goals.presence || "[]")
  rescue JSON::ParserError
    []
  end

  def goal_items=(items)
    self.goals = items.to_json
  end

  def status
    return "Unknown" if last_commit.nil?
    days = (Time.current - last_commit).to_f / 1.day
    return "Active"   if days <= 7
    return "Recent"   if days <= 30
    return "Stalled"  if days <= 90
    "Archived"
  end

  def time_ago
    return "never" if last_commit.nil?
    span = Time.current - last_commit
    minutes = span / 60
    hours   = span / 3600
    days    = span / 86400

    return "just now"                               if minutes < 1
    return "#{minutes.to_i}m ago"                   if hours < 1
    return "#{hours.to_i}h ago"                     if days < 1
    return "#{days.to_i}d ago"                      if days < 7
    weeks = (days / 7).to_i
    return "#{weeks} week#{'s' if weeks != 1} ago"  if days < 30
    months = (days / 30).to_i
    return "#{months} month#{'s' if months != 1} ago" if days < 365
    years = (days / 365).to_i
    "#{years} year#{'s' if years != 1} ago"
  end

  def tech_pills
    meta = metadata
    if meta["techs"].present?
      meta["techs"].split(",").map(&:strip).reject(&:blank?)
    elsif meta["inferred_type"].present? && meta["inferred_type"] != "Unknown"
      [meta["inferred_type"]]
    else
      []
    end
  end

  def contributor_count
    meta = metadata
    meta["contributors"].to_i
  end

  def total_commits
    meta = metadata
    meta["total_commits"].to_i
  end

  def branch_count
    meta = metadata
    meta["branch_count"].to_i
  end

  def doc_count
    meta = metadata
    meta["doc_count"].to_i
  end

  def contributor_names
    meta = metadata
    return [] if meta["contributor_names"].blank?
    meta["contributor_names"].split(",").map(&:strip)
  end

  def recent_commits
    meta = metadata
    return [] if meta["recent_commits"].blank?
    meta["recent_commits"].split(";;").filter_map do |entry|
      parts = entry.split("|", 3)
      date = Time.parse(parts[0]) rescue nil
      next nil unless date
      { date: date, author: parts[1] || "unknown", message: parts[2] || "" }
    end
  end

  def ownership_label
    is_forked ? "Fork" : "Owned"
  end

  def initial
    name.present? ? name[0].upcase : "?"
  end

  def status_color
    case status
    when "Active"  then "#34c759"
    when "Recent"  then "#0088ff"
    when "Stalled" then "#ffcc00"
    else "#8b8b90"
    end
  end

  def stall_risk
    case status
    when "Active"  then { label: "Low",      color: "#34c759", width: "15%" }
    when "Recent"  then { label: "Normal",   color: "#0088ff", width: "50%" }
    when "Stalled" then { label: "High",     color: "#ffcc00", width: "85%" }
    else                { label: "Archived", color: "#5b5b60", width: "100%" }
    end
  end
end
